rai/node/wallet.cpp

```diff
@@ -925,43 +929,73 @@ std::shared_ptr<rai::block> rai::wallet::change_action (rai::account const & sou
 	return block;
 }
 
-std::shared_ptr<rai::block> rai::wallet::send_action (rai::account const & source_a, rai::account const & account_a, rai::uint128_t const & amount_a, bool generate_work_a)
+std::shared_ptr<rai::block> rai::wallet::send_action (rai::account const & source_a, rai::account const & account_a, rai::uint128_t const & amount_a, bool generate_work_a, boost::optional<std::string> id_a)
 {
 	std::shared_ptr<rai::block> block;
+	bool cached_block = false;
 	{
-		rai::transaction transaction (store.environment, nullptr, false);
-		if (store.valid_password (transaction))
+		rai::transaction transaction (store.environment, nullptr, (bool)id_a);
+		if (auto id = id_a)
 		{
-			auto existing (store.find (transaction, source_a));
-			if (existing != store.end ())
+			for (auto i = rai::store_iterator (transaction, store.send_action_ids), n = rai::store_iterator (nullptr); i != n; ++i)
 			{
-				auto balance (node.ledger.account_balance (transaction, source_a));
-				if (!balance.is_zero () && balance >= amount_a)
+				if (i->first.size () == id->size () && memcmp (i->first.data (), id->data (), id->size ()) == 0)
 				{
-					rai::account_info info;
-					auto error1 (node.ledger.store.account_get (transaction, source_a, info));
-					assert (!error1);
-					rai::raw_key prv;
-					auto error2 (store.fetch (transaction, source_a, prv));
-					assert (!error2);
-					block.reset (new rai::send_block (info.head, account_a, balance - amount_a, prv, source_a, generate_work_a ? work_fetch (transaction, source_a, info.head) : 0));
+					rai::bufferstream stream (reinterpret_cast<uint8_t const *> (i->second.data ()), i->second.size ());
+					block = rai::deserialize_block (stream, rai::block_type::send);
```

Comment:
Do we need to store the whole block, or will the just the hash suffice?

If you can't lookup the block by hash to republish it, there's a good chance republishing it won't work anyways since it's probably a fork at that point (the send queue will have moved on, and another block will have taken it's place as the descendant of the parent block at the time).